### PR TITLE
Make sure to rebuild widgets even when they move

### PR DESCRIPTION
**Pull request by Hixie**
Previously in some rather esoteric cases involving global keys reversing
relative positions we would forget to build everyone.

### CODE REVIEW

File: packages/flutter/lib/src/widgets/framework.dart [hunk 1]
```diff
@@ -1487,6 +1487,25 @@ class BuildOwner {
 
   final List<BuildableElement> _dirtyElements = <BuildableElement>[];
   bool _scheduledFlushDirtyElements = false;
+  bool _dirtyElementsNeedsResorting; // null means we're not in a buildScope
+
+  /// Flags the dirty elements list as needing resorting.
+  ///
+  /// This should only be called while a [buildScope] is actively rebuilding the
+  /// widget tree.
+  void resortDirtyElements() {
```
Comment (abarth) [inline packages/flutter/lib/src/widgets/framework.dart]: markNeedsToResortDirtyElements?

File: packages/flutter/lib/src/widgets/framework.dart [hunk 2]
```diff
@@ -1641,6 +1682,7 @@ class BuildOwner {
       Timeline.finishSync();
       assert(_debugBuilding);
       assert(() {
+        _dirtyElementsNeedsResorting = null;
```
Comment (abarth) [inline packages/flutter/lib/src/widgets/framework.dart]: Why in an assert?  It seems dangerous for this field to have different values in production and debug given that it is used in both configurations.  I'd just assign null between lines 1681 and 1682.

**Review by abarth — APPROVED** [review]
LGTM

File: packages/flutter/lib/src/widgets/framework.dart (hunk 2 above)
Comment (Hixie) [inline packages/flutter/lib/src/widgets/framework.dart]: Micro-optimizing, I guess. I'll move it out.

**Pull request merged**